Gattc leak

jialu522 opened the pull request.

jialu522 committed:
bluetooth: Before releasing the IPC pending_queue, invoke the asynchronous callbacks for all messages contained within it.

bug: v/77564

When deleting the Bluetooth asynchronous instance, the IPC cached messages are directly released, and their asynchronous callbacks will not be invoked. This could lead to memory leaks if resources are released within the callbacks.

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: Release resources within the callback during deletion of the GATTC asynchronous instance.

bug: v/77564

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: In the asynchronous callback of GATTC deletion, release resources regardless of the operation status  (success or failure).

bug: v/77564

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: In GATTC discovery, report results as a single array per discovery event.

bug: v/77564

Report one service per discovery event. Data caching is required in this feature module. If the application exits during the discovery process, releasing the cached services becomes highly complex.

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: Check the validity of ins->context before use

bug: v/77564

When deleting the instance, ins->context is set to NULL. During asynchronous callback execution, ins->context may already be NULL.

Signed-off-by: jialu <jialu@xiaomi.com>

gzh-terry commented on feature/feature_async/src/bluetooth_ble_impl.c:
```diff
@@ -118,6 +118,10 @@ static bool gattc_userdata_type_cmp(void* node, void* type)
             break;                                                                               \
```
Can make the code shorter since the macro is used many times.
```
if (!ins || !ins->context) {
    ret = NULL;
    break;
}
features_info = (feature_bluetooth_features_info_t*)(ins->context); 
```

jialu522 replied:
ack

gzh-terry commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
+        return;
+
+    if (!packet) {
+        ret_cb(ins, BT_STATUS_UNHANDLED, 0, userdata);
```
`0` -> `BT_ADAPTER_STATE_OFF`

jialu522 replied:
ack

gzh-terry commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
+        return;
+
+    if (!packet) {
+        ret_cb(ins, BT_STATUS_UNHANDLED, 0, userdata);
```
`0` -> `BT_DEVICE_TYPE_UNKNOW`

jialu522 replied:
ack

gzh-terry commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
+        return;
+
+    if (!packet) {
+        ret_cb(ins, BT_STATUS_UNHANDLED, 0, userdata);
```
`0` -> `BT_SCAN_MODE_NONE`

jialu522 replied:
ack

gzh-terry commented on framework/socket/async/bt_device_async.c:
```diff
@@ -28,96 +28,180 @@ static void device_s8_reply(bt_instance_t* ins, bt_message_packet_t* packet, voi
+        return;
+
+    if (!packet) {
+        ret_cb(ins, BT_STATUS_UNHANDLED, 0, userdata);
```
`0` -> `BT_DEVICE_TYPE_UNKNOW`

jialu522 replied:
ack

huangyulong3 commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
 
-    if (ret_cb)
-        ret_cb(ins, packet->adpt_r.status, &packet->adpt_pl._bt_adapter_get_address.addr, userdata);
+    if (!ret_cb)
```
This code appears entirely redundant. Could we wrap it in a macro like the example below for easier future modifications?
```
#define HANDLE_BT_STATUS(ret_cb, packet, ins, userdata)   \
    if (!ret_cb) {                                          \
        return;                                             \
    }                                                        \
    if (!packet) {                                          \
        ret_cb(ins, BT_STATUS_UNHANDLED, NULL, userdata);   \
        return;                                             \
    }                                                        \
    ret_cb(ins, packet->adpt_r.status, &packet->adpt_pl._bt_adapter_get_address.addr, userdata);
```

jialu522 replied:
#define HANDLE_BT_ASTNC_CALLBACK(ret_cb, ins, packet, type, ...) \
    do {                                                         \
        if (!ret_cb)                                             \
            return;                                              \
        if (!packet) {                                           \
            ret_cb(ins, BT_STATUS_UNHANDLED, ##__VA_ARGS__);     \
            return;                                              \
        }                                                        \
        ret_cb(ins, packet->type##.status, ##__VA_ARGS__);       \
    } while (0)

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
 
-    if (ret_cb)
-        ret_cb(ins, packet->adpt_r.status, packet->adpt_pl._bt_adapter_get_name.name, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
 
-    if (ret_cb)
-        ret_cb(ins, packet->adpt_r.status, packet->adpt_pl._bt_adapter_get_uuids.uuids, packet->adpt_pl._bt_adapter_get_uuids.size, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
 
-    if (ret_cb)
-        ret_cb(ins, packet->adpt_r.status, packet->adpt_r.mode, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
 
-    if (ret_cb)
-        ret_cb(ins, packet->adpt_r.status, packet->adpt_r.ioc, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_adapter_async.c:
```diff
@@ -30,124 +30,226 @@ static void adapter_status_reply(bt_instance_t* ins, bt_message_packet_t* packet
-    if (ret_cb) {
-        ret_cb(ins, packet->adpt_r.status, packet->adpt_pl._bt_adapter_get_bonded_devices.addr,
-            packet->adpt_pl._bt_adapter_get_bonded_devices.num, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

gzh-terry commented on framework/socket/async/bt_device_async.c:
```diff
@@ -28,96 +28,180 @@ static void device_s8_reply(bt_instance_t* ins, bt_message_packet_t* packet, voi
+        return;
+
+    if (!packet) {
+        ret_cb(ins, BT_STATUS_UNHANDLED, 0, userdata);
```
`0` -> `BOND_STATE_NONE`

jialu522 replied:
ack

huangyulong3 commented on framework/socket/async/bt_device_async.c:
```diff
@@ -28,96 +28,180 @@ static void device_s8_reply(bt_instance_t* ins, bt_message_packet_t* packet, voi
-    if (ret_cb)
-        ret_cb(ins, packet->devs_r.status, userdata);
+    if (!ret_cb)
+        return;
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_device_async.c:
```diff
@@ -28,96 +28,180 @@ static void device_s8_reply(bt_instance_t* ins, bt_message_packet_t* packet, voi
 
-    if (ret_cb)
-        ret_cb(ins, packet->devs_r.status, packet->devs_r.bbool, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_device_async.c:
```diff
@@ -28,96 +28,180 @@ static void device_s8_reply(bt_instance_t* ins, bt_message_packet_t* packet, voi
 
-    if (ret_cb)
-        ret_cb(ins, packet->devs_r.status, packet->devs_r.v16, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

huangyulong3 commented on framework/socket/async/bt_device_async.c:
```diff
@@ -28,96 +28,180 @@ static void device_s8_reply(bt_instance_t* ins, bt_message_packet_t* packet, voi
 
-    if (ret_cb)
-        ret_cb(ins, packet->devs_r.status, packet->devs_r.v32, userdata);
+    if (!ret_cb)
```
Same as above.

jialu522 replied:
The number and types of parameters in asynchronous callbacks vary, making it difficult to implement with macros.

gzh-terry commented on framework/socket/async/bt_le_advertiser_async.c:
```diff
@@ -47,26 +52,44 @@ static void le_advertiser_bool_reply(bt_instance_t* ins, bt_message_packet_t* pa
+error:
+    data->adv = NULL;
+    free(adv);
+    ret_cb(ins, status, data->adv, data->userdata);
```
`ret_cb(ins, status, NULL, data->userdata);`

jialu522 replied:
ack

gzh-terry commented on framework/socket/async/bt_gattc_async.c:
```diff
@@ -109,17 +114,35 @@ static void gattc_create_connect_reply(bt_instance_t* ins, bt_message_packet_t*
+    bt_socket_async_client_t* priv = ins->priv;
     bt_gattc_delete_connect_cb_t ret_cb = (bt_gattc_delete_connect_cb_t)cb;
+    bt_gattc_remote_t* gattc_remote = (bt_gattc_remote_t*)data->gattc_remote;
+    void** user_phandle;
```
Remove extra line feed.
```
void** user_phandle = gattc_remote->user_phandle;
```

jialu522 replied:
ack

chengkai15 commented on feature/feature_async/src/bluetooth_ble_impl.c:
```diff
@@ -1275,68 +1274,55 @@ static void discover_callback(bt_instance_t* ins, gatt_status_t status, gattc_ha
 
-    feature_service->includeServices = include_service_array;
+    FeatureFreeValue(feature_service_array);
+    bt_list_remove(gattc_info->userdata_list, data);
```
why we need to remove userdata_list？

jialu522 replied:
This is intended to remove data from the userdata_list.

chengkai15 commented on framework/socket/async/bt_gattc_async.c:
```diff
@@ -109,17 +114,35 @@ static void gattc_create_connect_reply(bt_instance_t* ins, bt_message_packet_t*
+
+    user_phandle = gattc_remote->user_phandle;
+    bt_list_remove(priv->gattc_remote_list, gattc_remote);
+    *user_phandle = NULL;
```
where free *user_phandle？

jialu522 replied:
The user_phandle is a pointer to a pointer passed from the application.
user_phandle = gattc_remote.
In the GATT client (gattc client), you only need to remove the gattc_remote.
The application's data user_phandle does not require free().
Additionally, the application's user_phandle is merely a pointer (i.e., no memory was allocated via malloc), so it also does not require free().

jialu522 committed:
bluetooth: Using macros to reduce boilerplate code in asynchronous callbacks.

bug: v/77564

Some processing logic in asynchronous callbacks is identical; extract this common code into a macro.

Signed-off-by: jialu <jialu@xiaomi.com>

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

gzh-terry commented on framework/include/bt_async.h:
```diff
@@ -23,6 +23,17 @@ extern "C" {
 #include "bluetooth.h"
 
 #ifdef CONFIG_BLUETOOTH_FRAMEWORK_ASYNC
+#define HANDLE_BT_ASTNC_CALLBACK(ret_cb, ins, packet, type, userdata) \
```
Typo `ASYNC`

zhongzhijie1 commented on feature/feature_async/src/bluetooth_ble_impl.c:
```diff
@@ -1275,68 +1274,55 @@ static void discover_callback(bt_instance_t* ins, gatt_status_t status, gattc_ha
+    feature_service_array->_size = count;
+    feature_service_array->_element = calloc(feature_service_array->_size, sizeof(system_bluetooth_ble_GattService*));
+
     // service == NULL indicates the end of reporting
```
you may judge service == NULL and status == GATT_STATUS_SUCESS together, otherwise means discovery procedure is failure.

zhongzhijie1 reviewed (approved):
LGTM,

mind micro typo and service_searched_cb's status judge.
